Add a document describing the React I18n schema (#1599)
Shopify/quilt: packages/react-i18n/docs/react_i18n_schema.md

# Introduction <!-- omit in toc -->

React I18n is a schema comprised of a structure of nested key-value pairs (KVPs). This document captures the nuances of the constraints placed on the schema by the different levels of concern.

<!-- Created using "Markdown All in One" extension for VS Code -->
- [History](#history)
  - [Version 1.0](#version-10)
- [Versioning](#versioning)
- [Basic Structure](#basic-structure)
- [Pluralization](#pluralization)
  - [Reserved pluralization keys](#reserved-pluralization-keys)
  - [Invalid pluralization keys](#invalid-pluralization-keys)
  - [Detecting pluralization contexts](#detecting-pluralization-contexts)
- [Serialization](#serialization)
  - [Serialization to JSON](#serialization-to-json)
    - [Comments](#comments)
  - [Serialization to JSON+Comments](#serialization-to-jsoncomments)
- [Comparison with Rails I18n](#comparison-with-rails-i18n)
    - [Root locale key](#root-locale-key)
    - [Pluralization](#pluralization-1)
    - [Interpolation](#interpolation)

# History

## Version 1.0

> React I18n's schema is loosely based on the [Rails I18n](https://guides.rubyonrails.org/i18n.html) schema.

Reviewer: Is this necessary with the comparison section in the TOC? Since that section starts with the same sentence.

Author: Removed.

Prior to the creation of this document, the React I18n schema was not documented.
With the creation of this document, the schema was specified and given the version identifier `1.0`.

# Versioning

The version identifier for the schema is made of two parts, `<major>` and `<minor>`, separated by a `.`: `<major>.<minor>`.

> the `<major>` version number will be incremented any time a change is introduced that would mean that existing parsers would break while parsing the new schema. Put another way, software that supports parsing version `x.y` of the schema will always be able to parse any version `x.z`, where `z >= y`.

Reviewer: ```suggestion
The `<major>` version number will be incremented any time a change is introduced that would mean that existing parsers would break while parsing the new schema. Put another way, software that supports parsing version `x.y` of the schema will always be able to parse any version `x.z`, where `z >= y`.
```

> the `<major>` version number will be incremented any time a change is introduced that would mean that existing parsers would break while parsing the new schema. Put another way, software that supports parsing version `x.y` of the schema will always be able to parse any version `x.z`, where `z >= y`.

Reviewer: > Put another way, software that supports parsing version `x.y` of the schema will always be able to parse any version `x.z`, where `z >= y`.

True, but should these two paragraphs be re-organized so that this is at the end? This example also demonstrates the meaning of a `<minor>` version change.

Author: 🤷 ... I'm going to keep it this way.

# Basic Structure

React I18n is made of nested dictionaries of key-value pairs (KVPs). Each of the keys are strings, and the values are either strings, or a nested dictionary.

In TypeScript, the type is a recursive definition:

```typescript
export interface TranslationDictionary {
  [key: string]: string | TranslationDictionary;
}
```

When [serialized to JSON]](#serialization-to-json), each of the dictionaries become JSON `object` types, and the keys and leaf values are both JSON `string` types.

**Example React I18n JSON file**

```jsonc
{
  "hello": "My name is {name}",
  "parent": {
    "child_1": "I am the first child of `parent`!",
    "child_2": {
      "grandchild_1": "I am the first grandchild of `parent`!",
      "grandchild_2": "I am the second grandchild of `parent`!",
    }
  }
}
```

# Pluralization

Pluralization is handled by providing the pluarlization keys required by the language as leaf KVPs, creating a "pluralization context".

**Example English file with pluralization keys:**
```jsonc
{
  "cars": {  // All the children of `cars` are in a "pluralization context"
    "one": "I have {count} car",
    "other": "I have {count} cars"
  }
}
```

> Different languages require different sets of pluralization keys. For example, the Polish (`pl`) has 4 required pluralization keys: `one`, `few`, `many`, and `other`. The Polish version of the previous example would be:

Reviewer: ```suggestion
Different languages require different sets of pluralization keys. For example, Polish (`pl`) has 4 required pluralization keys: `one`, `few`, `many`, and `other`. The Polish version of the previous example would be:
```
or
```suggestion
Different languages require different sets of pluralization keys. For example, the Polish language (`pl`) has 4 required pluralization keys: `one`, `few`, `many`, and `other`. The Polish version of the previous example would be:
```

**Example Polish file with pluralization keys:**
```jsonc
{
  "cars": {
    "one": "Mam {count} samochód",
    "few": "Mam {count} samochody",
    "many": "Mam {count} samochodów",
    "other": "Mam {count} samochodu",
  }
}
```

## Reserved pluralization keys

> In order to avoid collisions, the following keys are reserved for use as pluralization keys, and must not be used as keys of leaf KVPs outside of the pluralization context:

Reviewer: ```suggestion
In order to avoid name collisions, the following keys are reserved for use as pluralization keys and must not be used as keys of leaf KVPs outside of a pluralization context:
```

Reviewer: Can I use reserved key names for internal nodes?

Author: > Can I use reserved key names for internal (non-leaf) nodes?

@sjahu, yes. AFAIK, there is no need to restrict usage in non-leaf nodes.

Author: Fixed.

* `few`
* `many`
* `one`
* `other`
* `two`
* `zero`

These keys are derived from the names given to the pluralization rules of all languages, as defined in the [Unicode Consortium's Common Locale Data Repository (CLDR)](https://github.com/unicode-org/cldr).

This list is current as of verison `37` of the CLDR. In the unlikely event that new pluralization rules names are added, they will be added to this reserved list in a future version of this specification.

**Example INVALID English file:**
```jsonc
{
  "foo": "bar",
  "other": "Other" // INVALID: `other` is a keyword reserved for use in pluralization contexts!
}
```

## Invalid pluralization keys

> Further, pluralization keys that are not required by the language must not be used, even within a pluralization context.

Reviewer: ```suggestion
Furthermore, pluralization keys that are not required by the language must not be used, even within a pluralization context.
```

> For example, even though `zero` is a required pluralization key for some languages (eg. Arabic), it is not a required pluralization key in English. Therefore, the `zero` key should not be present in an English file alongside the pluralization KVPs required by English.

Reviewer: ```suggestion
For example, even though `zero` is a required pluralization key for some languages (eg. Arabic), it is not a required pluralization key in English. Therefore, the `zero` key must not be present in an English file alongside the pluralization KVPs required by English.
```

**Example INVALID English file:**
```jsonc
{
  "cars": {
    "zero": "I have no cars", // INVALID: `zero` is not a required pluralization key for English!
    "one": "I have {count} car",
    "other": "I have {count} cars"
  }
}
```

## Detecting pluralization contexts

These constraints allow for an algorithm for identifying pluralization contexts:

> If a node has child leaf KVPs for each of the pluralization keys required by the language, then its children are within a pluralization context.

Reviewer: True, but could this be strengthened to simply "If a node has child leaf KVPs for any of the pluralization keys, then ..."? Since pluralization keys outside this context are invalid?

Author: You are right. Fixed.

```jsonc
{
  // This is a context comment for `hello`!
  "hello": "My name is {name}",
  "parent": {
    "child_1": "I am the first child of `parent`!", // This is a trailing context comment for `child_1`!
    "child_2": {
      /*
        This is a multi-line
        context comment for `grandchild_1`!
      */
      "grandchild_1": "I am the first grandchild of `parent`!",
      "grandchild_2": "I am the second grandchild of `parent`!" /* This is a trailing context comment for `grandchild_2`! */
    }
  }
}
```

Comments follow the [JSON5 syntax for comments](https://spec.json5.org/#comments): `//` signals the start of a comment that ends at the end of the line. `/*` signals the start of a comment that ends with the next `*/`.

# Comparison with Rails I18n

React I18n's schema is loosely based on the [Rails I18n](https://guides.rubyonrails.org/i18n.html) schema.
This section compares the two schemas, in order to point out the ways in which they differ from one another.

### Root locale key

Rails I18n has a locale root key, while React I18n doesn't:

**Rails I18n:**
```yaml
en:
  foo: bar
```

**React I18n**

```json
{
  "foo": "bar"
}
```

> Note the `en` (for English) locale key that is the root key for the Rails I18n file. This key is not present in the React I18n file because it is denoted by filename.

Reviewer: > denoted by filename

If this part of the specification it should be more explicit. If not, we should add a sentence stating that React I18n files are locale-agnostic (not true, though, since the set of valid pluralization keys is determined by the locale).

Author: Changed to make it clear that the locale doesn't have to be part of the filename, but typically is.

### Pluralization

Rails I18n (when using the default backend) allows for the use of a `zero` key (regardless of the source language), while React I18n doesn't.

The use of `zero` in Rails I18n is a problematic, as there are languages (eg. Arabic) that use `zero` as a pluralization key.
This causes collisions when `zero` is supplied in a source file for a language that doesn't use the `zero` pluralization key (eg. English).

React I18n avoids this by disallowing the use of the `zero` key unless the source language requires the `zero` pluralization key.

When manually converting Rails I18n files to React I18n, use a separate key for storing strings that should be used for placeholder or empty values.

**Rails I18n:**
```yaml
en:
  cars:
    zero: "I don't have any cars" # Problematic when translating into languages that use the `zero` pluralization key
    one: "I have %{count} car"
    other: "I have %{count} car"
```

**React I18n**

```json
{
  "cars": {
    "one": "I have {count} car",
    "other": "I have {count} cars"
  },
  "no_cars": "I don't have any cars"
}
```

### Interpolation

> While neither schemas define an interpolation syntax, the most common syntax used in Rails I18n is `%{}`, while in React I18n, the most common syntax is `{}`.

Reviewer: ```suggestion
While neither schema defines an interpolation syntax, the most common syntax used in Rails I18n is `%{}`, while in React I18n, the most common syntax is `{}`.
```

> # Introduction <!-- omit in toc -->

Reviewer 2: Seems to omit the section from the ToC:
![image](https://user-images.githubusercontent.com/50832/94567760-cebfcd80-0239-11eb-9029-cb81ebff059b.png)

> # Introduction <!-- omit in toc -->

Author: Yes, that is the syntax that [Markdown All In One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one) uses to prevent the heading from getting added to the auto-generated table of content.

React I18n is a schema comprised of a structure of nested key-value pairs (KVPs). This document captures the nuances of the constraints placed on the schema by the different levels of concern.

<!-- Created using "Markdown All in One" extension for VS Code -->
- [History](#history)
  - [Version 1.0](#version-10)
- [Versioning](#versioning)
- [Basic Structure](#basic-structure)
- [Pluralization](#pluralization)
  - [Reserved pluralization keys](#reserved-pluralization-keys)
  - [Invalid pluralization keys](#invalid-pluralization-keys)
  - [Detecting pluralization contexts](#detecting-pluralization-contexts)
- [Serialization](#serialization)
  - [Serialization to JSON](#serialization-to-json)
    - [Comments](#comments)
  - [Serialization to JSON+Comments](#serialization-to-jsoncomments)
    - [Why not JSON5?](#why-not-json5)
- [Comparison with Rails I18n](#comparison-with-rails-i18n)
    - [Root locale key](#root-locale-key)
    - [Pluralization](#pluralization-1)
    - [Interpolation](#interpolation)

# History

## Version 1.0

> React I18n's schema was loosely based on the Rails I18n schema.

Author: Done.

Prior to the creation of this document, the React I18n schema was not documented.

> With the creation of this document, the schema was specified and given the version identifier `1.0`.

Author: The `@shopify/react-i18n` library changes independently of the React I18n specification. So they have different version numbers.

React I18n is the specification, while `@shopify/react-i18n` is a client implementation that supports files in the React I18n schema.

This is the same as any other schema / format specification. For example, YAML is [version 1.2](https://yaml.org/spec/1.2/spec.html), but each implementation library has their own version numbers.

> With the creation of this document, the schema was specified and given the version identifier `1.0`.

Reviewer 3: If we define the "React I18n specification" and `@shopify/react-i18n` is the only library parsing a target schema version, I still don't see the reason for distinction. I might be the only person who thinks this, but I'd like to see what others think.

> With the creation of this document, the schema was specified and given the version identifier `1.0`.

Reviewer 3: > @shopify/react-i18n isn't the only parser/client though.

I see, sorry I didn't realize that. Separate versioning seems good to me under those circumstances 👍

}
```

## Invalid pluralization keys

Further, pluralization keys that are not required by the language must not be used, even within a pluralization context.

For example, even though `zero` is a required pluralization key for some languages (eg. Arabic), it is not a required pluralization key in English. Therefore, the `zero` key should not be present in an English file alongside the pluralization KVPs required by English.

**Example INVALID English file:**
```json5
{
  "cars": {
    "zero": "I have no cars", // INVALID: `zero` is not a required pluralization key for English!
    "one": "I have {count} car",
    "other": "I have {count} cars"
  }
}
```

## Detecting pluralization contexts

These constraints allow for an algorithm for identifying pluralization contexts:

If a node has child leaf KVPs for each of the pluralization keys required by the language, then its children are within a pluralization context.

Any additional children within the context are invalid.

# Serialization

React I18n can be serialized to different formats. However, not all serialization formats are created equal.

> Two serializations are officialy defined. Consumers of React I18n `1.0` files are expected to be able to parse files serialized into each of these serialization formats.

Reviewer 3: Ambiguous version here.

Author: Changed to `Consumers of React I18n `1.0` schema files` (added `schema`).

| Serialization Format | JSON+Comments | JSON |
| -------------------- | ------------- | ---- |
| Comments             | ✅             | ❌    |

> ## Serialization to JSON

Reviewer 3: I don't think this is worth repeating. Can we remove it?

Author: Done.

```

Historically, `JSON` had support for comments, but this [was then removed](https://archive.vn/20150704102718/https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr). However, comments have proven to be a useful mechanism for developers to [provide additional context to translators](https://development.shopify.io/engineering/developing_at_Shopify/internationalization/providing_context_notes). The `JSON+Comments` serialization format extends the `JSON` serialization to re-add support for comments.

The `JSON+Comments` can be parsed using a `JSON5` parser, or a `JSON` parser that is not sensitive to comments.

**Example React I18n JSON+Commments file**

```json5
{
  // This is a context comment for `hello`!
  "hello": "My name is {name}",
  "parent": {
    "child_1": "I am the first child of `parent`!", // This is a trailing context comment for `child_1`!
    "child_2": {
      /*
        This is a multi-line
        context comment for `grandchild_1`!
      */
      "grandchild_1": "I am the first grandchild of `parent`!",
      "grandchild_2": "I am the second grandchild of `parent`!" /* This is a trailing context comment for `grandchild_2`! */
    }
  }
}
```

> Comments follow the [JSON5 syntax for comments](https://spec.json5.org/#comments): `//` signals the start of a comment that ends at the end of the line. `/*` signals the start of a comment that ends with the next `*/`.

Reviewer 3: This tangent about JSON5 and comments is a bit irrelevant to the schema definition don't you think? Can we move to another doc that talks about serialization for those interested and link it here?

Author: Removed.

### Why not JSON5?

Ideally, we would not have needed a `JSON+Comments` serialization, and instead just defined a [`JSON5`](https://json5.org/) serialization instead.
However, `JSON5` parsers are immature, and notably there are no mature `JSON5` parsers in Ruby. Given that important consumers of React I18n files are written in Ruby, it would not make sense to define a serialization using the full syntax of `JSON5`, since those files would not be able to be handled.

Adding a `JSON5` serialization would be an excellent future addition to React I18n, at a point in time where these consumers could be made to understand `JSON5`.

Since `JSON+Comments` is a subset of `JSON5`, a `JSON5` parser can be used to parse `JSON+Comments` serialized files.

# Comparison with Rails I18n

React I18n's schema was loosely based on the Rails I18n schema.
This section compares the two schemas, in order to point out the ways in which they differ from one another.

### Root locale key

Rails I18n has a locale root key, while React I18n doesn't:

**Rails I18n:**
```yaml
en:
  foo: bar
```

**React I18n**

```json
{
  "foo": "bar"
}
```

> Note the `en` (for English) locale key that is the root key for the Rails I18n file. This key is not present in the React I18n file.

Reviewer 3: ```suggestion
Note the `en` (for English) locale key that is the root key for the Rails I18n file. This key is not present in the React I18n file because it is denoted by filename.
```
